### PR TITLE
Hide send message button in group chat if user tries to spam

### LINKED ISSUE #4474: Add a timeout on spamming user's send button
[comment]: # (Please replace ... with your information. Remove < and >)

### User Story

As a user, I want to have the send button of a spamming user's app be disabled for a certain period after each sent message so that spamming a chat is less feasible for him.

### Description

[comment]: # (Feature or Bug? i.e Type: Bug)
*Type*: Feature

[comment]: # (Describe the feature you would like, or briefly summarise the bug and what you did, what you expected to happen, and what actually happens. Sections below)
*Summary*: Currently, it is very easy for a user to spam a public chat by pasting some long message over and over again. If we disable the send button for an increasing amount of time each time that a user sends messages in quick succession, we can discourage that behavior.

#### Expected behavior
[comment]: # (Describe what you expected to happen.)
- User sends message 1
- User sends message 2 within 1 second
- App notices that the user is sending messages in quick succession and increases the timeout during which the Send button is disabled to 2 seconds.
- User sends another message right after the button is enabled again.
- App increases timeout to 5 seconds.

Note: There is a cooldown period of 10 seconds, after which we reset the timeout. These values should be easy to change so we can experiment with them.

#### Actual behavior
[comment]: # (Describe what actually happened.)
- User is able to send messages in quick succession without any limits.

### PR DESCRIPTION
**Pull request by dmitryn**
fixes #4474 

### Summary:

Hides send message button in group chat if user tries to spam.

Spam is considered spam when interval between current and last message sent by user is less than 1 second. When it happens, send button gets hidden for 2 seconds. If user spams again, cooldown gets increased to 5 seconds. Third cooldown gets raised to 10 seconds.

### Steps to test:
- Open Status
- Join public channel
- Send message
- Send one more message in less than 1 second
- Type something
- Check that send button is hidden

status: ready

### CODE REVIEW

**Comment by rasom** [comment]
apk uploaded to https://i.diawi.com/htTJjY for branch PR-4537

<blockquote><div><strong><a href="https://i.diawi.com/htTJjY">Install - Diawi - Development and In-house Apps Wireless Installation</a></strong></div></blockquote>

**Comment by rasom** [comment]
ipa uploaded to https://i.diawi.com/iB4yk1 for branch PR-4537

<blockquote><div><strong><a href="https://i.diawi.com/iB4yk1">Install - Diawi - Development and In-house Apps Wireless Installation</a></strong></div></blockquote>

**Review by janherich — APPROVED** [review]

**Comment by pedropombeiro** [comment]
@dmitryn I just sent a few messages to #status in quick succession and didn't see any changed behavior (I did build locally for Android instead of using the apk, but that should work)., Any ideas?

UPDATE: It's weird, I created a #Test public channel and there it worked as expected (kind of). Sometimes I'm able to keep on sending new messages without the logic kicking in (after 1-2 successful restrictions).

**Comment by dmitryn** [comment]
@PombeirP if button is still there, maybe your intervals between spamming not < 1s. I can set it 2s so it will be easier to test against.

**Comment by dmitryn** [comment]
Or i can even move that parameter to config so you could make custom build with it.

**Comment by pedropombeiro** [comment]
I have the impression that we'd want `spam-interval-ms` to be an array, so that your message is considered to be spam at first if within 1s, then if within 2s, etc.

**Comment by dmitryn** [comment]
Yes, we could gradually increase `spam-interval-ms`. The issue is that spammer knowing our threshold, can just send message every N+1 second without any cooldown.

upd: where N is `spam-interval-ms`

**Comment by pedropombeiro** [comment]
Yeah, but if we're gradually increasing `spam-interval-ms`, we should be already limiting the harm he can do. So the cooldown period of 10 secs would only count after a message was sent which was not considered to be spam (i.e. cooldown is reached if time of current message > time of previous message + spam-interval-ms + cooldown-ms). Does that make sense?

**Comment by dmitryn** [comment]
>Yeah, but if we're gradually increasing spam-interval-ms, we should be already limiting the harm he can do.

Agree, i will add gradual increase.

>So the cooldown period of 10 secs would only count after a message was sent which was not considered to be spam (i.e. if time of current message > time of previous message + spam-interval-ms + cooldown-ms).

Not sure i understand this. Reset `spam-interval-ms` back to 1s when user starts messaging on normal basis (not spamming) ?

**Comment by pedropombeiro** [comment]
> Not sure i understand this. Reset spam-interval-ms back to 1s when user starts messaging on normal basis (not spamming) ?

Exactly. So if the user is past `spam-interval-ms` and waits for 10 more seconds (the cooldown period), then `spam-interval-ms` is reset.

**Comment by rasom** [comment]
apk uploaded to https://i.diawi.com/kdpe5d for branch PR-4537

<blockquote><div><strong><a href="https://i.diawi.com/kdpe5d">Install - Diawi - Development and In-house Apps Wireless Installation</a></strong></div></blockquote>

**Comment by rasom** [comment]
ipa uploaded to https://i.diawi.com/nPGXtZ for branch PR-4537

<blockquote><div><strong><a href="https://i.diawi.com/nPGXtZ">Install - Diawi - Development and In-house Apps Wireless Installation</a></strong></div></blockquote>

**Comment by dmitryn** [comment]
@PombeirP made the updates to current PR.

@yenda @cammellos please review.

File: src/status_im/chat/models/input.cljs [hunk 1]
```diff
@@ -198,3 +199,24 @@
...
+                     :ms       (const/cooldown-periods-ms cooldowns
+                                                          const/default-cooldown-period-ms)}]
+   :db             (assoc db
+                          :chat/cooldowns (if (= const/cooldown-reset-threshold cooldowns)
+                                            0
+                                            cooldowns)
+                          :chat/send-button-disabled? true)})
+
+(defn process-cooldown [{{:keys [chat/last-outgoing-message-sent-at chat/cooldowns current-chat-id] :as db} :db}]
+  (when (get-in db [:chats current-chat-id :group-chat])
```
Comment (cammellos) [inline src/status_im/chat/models/input.cljs]: You could reuse:

https://github.com/status-im/status-react/blob/7ef3a5a5dab34500608ef30f3e2dd4b9ce95f55a/src/status_im/chat/models.cljs#L12

`(models.chat/multi-user-chat? current-chat-id cofx)`

hopefully there's no circular dependency :)

File: test/cljs/status_im/test/chat/models/input.cljs [hunk 2]
```diff
@@ -166,3 +168,53 @@
...
+                                               :chat/cooldowns 1
+                                               :chat/send-button-disabled? true)
+                        :dispatch-later [{:dispatch [:enable-send-button]
+                                          :ms       (constants/cooldown-periods-ms 1)}]}
+              actual (input/process-cooldown {:db db})]
+          (is (= expected actual))))
+
+      (testing "spamming detected twice"
+        (let [db (assoc db :chat/cooldowns 1
+                        :chat/last-outgoing-message-sent-at (- 1527675198542 900))
```
Comment (cammellos) [inline test/cljs/status_im/test/chat/models/input.cljs]: indentation is a bit off

**Review by cammellos — APPROVED** [review]
Looks good to me!

**Comment by asemiankevich** [comment]
@hesterbruikman @andmironov @jpbowen i would like to ask you for review of  this PR which hides `send` when i send messages faster then 1 per second in  public and group chats. I see 2 concerns here. 1 - i personally do this all the time with no intention to spam anybody 2 - this is very easy to avoid if someone wants to spam the channel really. I am not sure this realization makes sense in current state. Please take a look! :raised_hands: (edited)

**Comment by asemiankevich** [comment]
It was agreed to make 3+ messages sent in quick seccession should be considered as a spam (adjusted by parameter)

**Comment by dmitryn** [comment]
>It was agreed to make 3+ messages sent in quick seccession should be considered as a spam (adjusted by parameter)

@asemiankevich added this behavior, please check.

Also, flag `SPAM_BUTTON_DETECTION_ENABLED` added which is enabled for dev and nightly only.

**Comment by asemiankevich** [comment]
@dmitryn was this build triggered with this flag enabled ? seems it is not working now at all , nothing is considered as spam in current build ^

**Comment by dmitryn** [comment]
@asemiankevich sorry, configuration flag was not set. Waiting for new build to be built..

**Comment by asemiankevich** [comment]
thanks @dmitryn ! seems it is working in public chats, but not in group. Is this on purpose?

**Comment by andmironov** [comment]
hey, sorry if it's too late, share @asemiankevich 's concern, i think it might be considered as a bug by many users. i think there should be a more delicate way to prevent spam: maybe this should happen after a user get's reported by someone and even after that, there should be a way to prove, that you are not a spammer and make the Send button visible all the time.

So i would say that this requires some more research to be done

**Comment by dmitryn** [comment]
>seems it is working in public chats, but not in group. Is this on purpose?

@asemiankevich yes, should only happen in public chats.

**Comment by dmitryn** [comment]
@andmironov instead of hiding button, we could show it as inaccessible (maybe different color) and display text about cooldown in input field so user would be more aware of what's happening.

**Comment by andmironov** [comment]
as a quick fix i would suggest doing something like this:
1. Inform the user what happened
2. Add a timeout to make it clear when sending messages is going work

but this is just a quick fix, far from perfect

bl2

**Comment by asemiankevich** [comment]
@andmironov @dmitryn what is the decision?

**Comment by andmironov** [comment]
@asemiankevich if this is critical and we don't have time for a more elegant/technically sophisticated approach (or if we consider this approach elegant enough, which i am not 100% sure about :-) ), i can prep up a mockup doc and upload this to Zeplin for both IOS and Android. 

cc @chadyj (how critical is it?) @jpbowen (is this good enough? could you think of any bad flows because of this?)

**Comment by andmironov** [comment]
questions so far:
1. What exactly should be done by the user to get blocked? (sending messages with a small interval? sending big messages? getting reported by someone?)
2. What should we say in the message (`Sorry, sending messages in this chat is temporary blocked for you`)?
3. How long should the timer go? (1 min?)
4. What happens if the user gets blocked the second time, does the timer go longer?
5. What if user is not a spammer and gets blocked and has to send an urgent message? We should be really clear on how we track spammers to help users not to get blocked.

So i think, considering the above, in general this is not ready to be pushed to the app.

**Comment by chadyj** [comment]
The spam issue is real, but as @andmironov mentioned there is a better solution out there. Let's put our heads together and figure this out.

But since this PR is done I'd suggest the following changes:

* Relax the restrictions slightly. I'd look at something like restricting more than 3 messages in 5 secs. In  normal chat usage it is pretty normal to send a few. messages. in. rapid. succession. We don't want regular users encountering this restriction.
* Don't hide the UI or change UI states. This is anti-user and confusing if non-spammy users encounter it. Instead, keep the UI, but pop an alert saying something like "Sorry, we limit sending several messages in quick succession to prevent spam. Please try again in a moment". This explains the issue, why we do it, and lets legit users know how to continue sending.
* Start thinking about comprehensive spam tools.

**Comment by hesterbruikman** [comment]
@PombeirP please correct me if I'm wrong.

This issue was also discussed in the last #core-chat sync, Mayn29. I think we all agreed that blocking the send button is indeed not the most elegant solution nor the most desirable one. While working on implementing this shortterm, I believe the team was going to investigate alternatives that have been discussed and don't block usage.

- [ ] Applying the collapse mechanisms of 1:1 to public chat.
- [ ] Bundling messages of a sender as they come in and present them with a counter and expand option.
- [ ] Warning the listener if spam is detected to more easily enable to 'mute' the channel

The second being most sophisticated, but also the most complex (both technically and ethically) as it requires 'filtering' by sender.

**Comment by asemiankevich** [comment]
I am returning this to backlog until the decision is made. I would not merge this right away in current implementation

**Comment by pedropombeiro** [comment]
@hesterbruikman that is correct. Regarding the first item, @janherich [mentioned](https://status-im.slack.com/archives/C9B1T4B9B/p1527853517000015?thread_ts=1527848536.000102&cid=C9B1T4B9B) it would be easy to collapse the messages, but it quickly became clear that the problem would lie in the UX (there are some good UX questions raised by him [there](https://status-im.slack.com/archives/C9B1T4B9B/p1527854061000272?thread_ts=1527848536.000102&cid=C9B1T4B9B)). On that thread, the agreement seemed to be that we just needed to relax the constraints a little bit for this PR, so I don't quite understand why we're abandoning it (unless we decide we prefer to allow spam on the beta release, because we don't seem to have any short-term alternative).

**Comment by pedropombeiro** [comment]
@oskarth is asking whether this is still a beta blocker. Should we merge this for beta with the relaxed restrictions and a popup showing up when spam condition is triggered? Please vote :+1:  or :-1:

**Comment by dmitryn** [comment]
Updates made to current PR:
* Relaxed restrictions on spam detection rules: 5 successive messages sent with <1 second interval considered as a spam in public chat.
* Show warning message `"Sorry, we limit sending several messages in quick succession to prevent spam. Please try again in a moment"`
* Disable text input during cooldown. I didn't find a way to prevent user from closing warning dialog as no corresponding api option exist for RN `Alert` component, so user can close warning message dialog and continue spamming.
* Change input placeholder to `"Please wait a moment..."` (wording is not perfect, but user already has context from warning message shown previously, so he knows why he needs to wait) 

screenshot 2018-06-13 21 00 15

**Comment by lukaszfryc** [comment]
@dmitryn the dialog on Android can be dismissed without clicking the "OK" button. It means that spammer can easily spam the chat without even noticing the dialog. If we can fix that, that would be great. On iOS, you need to click on OK to dismiss it. It stops spamming.

Otherwise works good. The dialog didn't appear when I used the app "in normal way".

**Comment by dmitryn** [comment]
@lukaszfryc i've made a change to prevent dialog from dismissing without clicking OK on Android. Please check.

**Comment by lukaszfryc** [comment]
Dialog works fine now. 

But, I noticed one more thing. Sometimes, the last message that was sent before the dialog appeared has the spinner and looks like it was not sent:



But, **the message actually was sent**, I can see it on another device. The spinner does not go away even after 10 min or killing the app and opening it again. 

@dmitryn it's minor but annoying since the spinner stays forever. Could you take a look at it? Try to send those 5 messages really quickly.

**Comment by dmitryn** [comment]
>Could you take a look at it? Try to send those 5 messages really quickly.

@lukaszfryc i cannot reproduce this error. Tried on android and ios simulators and on real android, but i'm using dev build. Let me try this one https://i.diawi.com/GNb7GB

<blockquote><div><strong><a href="https://i.diawi.com/GNb7GB">Install - Diawi - Development and In-house Apps Wireless Installation</a></strong></div></blockquote>

**Comment by lukaszfryc** [comment]
@dmitryn try to send the messages quickly, until the dialog appears. Try couple of times

**Comment by dmitryn** [comment]
@lukaszfryc still cannot reproduce. Does it happen only after dialog is shown? I could assume this bug is not related to current PR, maybe it makes sense to create separate issue for that.

**Comment by lukaszfryc** [comment]
@dmitryn I can't reproduce it 100%. Maybe it's related to quality of my network connection. I agree with you to create a separate issue for it.

**Pull request merged**